test/simulator/moves/encore.js

```diff
@@ -0,0 +1,124 @@
+'use strict';
+
+const assert = require('./../../assert');
+const common = require('./../../common');
+
+let battle;
+
+describe('Encore', function () {
+	afterEach(function () {
+		battle.destroy();
+	});
+
+	it('should not affect Focus Punch if the the user\'s decision is not changed', function () {
+		battle = common.createBattle({gameType: 'doubles'}, [
+			[
+				{species: "Smeargle", level: 50, ability: 'owntempo', moves: ['splash', 'focuspunch']},
+				{species: "Abra", level: 1, ability: 'innerfocus', moves: ['knockoff', 'teleport']},
+			],
+			[
+				{species: "Smeargle", ability: 'owntempo', item: 'laggingtail', moves: ['encore', 'splash']},
+				{species: "Zigzagoon", level: 1, ability: 'pickup', moves: ['extremespeed']},
+			],
+		]);
+
+		// If the Focus Punch user is not interrupted the attack is expected to be successful.
+		battle.makeChoices('move focuspunch 1, move knockoff 1', 'move splash, move extremespeed 2');
+		const hp = battle.p2.active[0].hp;
+		assert.notStrictEqual(hp, battle.p2.active[0].maxhp);
+
+		// If a user's previous move was Focus Punch and it is Encored into Focus Punch while attempting to
+		// execute the move, the regular "you can't be hit" effect for Focus Punch will be enforced.
+		battle.makeChoices('move focuspunch 1, move teleport', 'move encore 1, move extremespeed 1');
+		assert.strictEqual(hp, battle.p2.active[0].hp);
+
+		// During subsequent turns the normal Focus Punch behavior applies.
+		battle.makeChoices('move focuspunch 1, move teleport', 'move splash, move extremespeed 1');
+		assert.strictEqual(hp, battle.p2.active[0].hp);
+	});
+
+	it('should make Focus Punch always succeed if it changes the user\'s decision', function () {
+		battle = common.createBattle({gameType: 'doubles'}, [
+			[
+				{species: "Smeargle", level: 50, ability: 'owntempo', moves: ['splash', 'focuspunch']},
+				{species: "Abra", letvel: 1, ability: 'innerfocus', moves: ['knockoff', 'teleport']},
+			],
+			[
+				{species: "Smeargle", ability: 'owntempo', item: 'laggingtail', moves: ['encore', 'splash']},
+				{species: "Zigzagoon", level: 1, ability: 'pickup', moves: ['extremespeed']},
+			],
+		]);
+
+		// If the Focus Punch user is interrupted the attack is not expected to be successful.
+		battle.makeChoices('move focuspunch 1, move knockoff', 'move splash, move extremespeed 1');
+		let hp = battle.p2.active[0].hp;
+		assert.strictEqual(hp, battle.p2.active[0].maxhp);
+
+		// The Pokemon Encored into Focus Punch is not subject to the negative effects of Focus Punch; that is,
+		// if it is hit at any time before or after the Encore, it still uses Focus Punch like normal. It doesn't matter
+		// in the case of Focus Punch if the user was hit before or after the Encore; Focus Punch will still always work.
+		battle.makeChoices('move splash, move teleport', 'move encore 1, move extremespeed 1');
+		assert.notStrictEqual(hp, battle.p2.active[0].hp);
```

Comment:
You'll have to make the previous turn a successful Focus Punch to get this to work, sorry about that!

Reply:
> Sorry, I don't know why I said this; Focus Punch can't be Encored if it failed, but Shell Trap can. And they can both be Encored if they succeeded.

No worries, I'm just in awe that you guys know all this! 

> You'll have to make the previous turn a successful Focus Punch to get this to work, sorry about that!

Ah, OK. So changing this back would cause the test to pass, indicating Focus Punch is in fact implemented correctly in PS today. Is that right?

Reply:
Yeah, Focus Punch behaviour is currently correct as far as I know.